### PR TITLE
[ONNX] Diagnostic to show all unsupported call_functions

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100451
* #100407

Introduce `Analysis` to analyze fx graphmodule and emit diagnostics. This class
can be extended to interact with `Transform` (passes) to decide if a pass should
trigger based on graph analysis result. E.g., if decomp needs to run by checking
operator namespace in nodes. For now leaving it as out of scope but can revisit
if maintaining multi fx extractor becomes reality.

### CODE REVIEW

File: torch/onnx/_internal/fx/function_dispatcher.py [hunk 1]
```diff
@@ -306,3 +309,82 @@ def _create_onnx_friendly_decomposition_table() -> (
 DEFAULT_ONNX_EXPORTER_DECOMPOSITION_TABLE: Dict[
     torch._ops.OpOverload, Callable
 ] = _create_onnx_friendly_decomposition_table()
+
+
+def get_symbolic_function(
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/function_dispatcher.py:314]: note: this part is extracting dispatch code from `fx_to_onnxscript.py` to `function_dispatcher.py`. The goal is to make it re-usable for the analysis pass. A proper dispatcher refactor is still required, and is not the scope of this PR.

Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:314]: #100660 targets the refactor of `function_dispatcher`

File: torch/onnx/_internal/fx/_pass.py [hunk 2]
```diff
@@ -204,3 +204,22 @@ def run(self, *args, **kwargs) -> torch.fx.GraphModule:
...
 
         return module
+
+
+class AnalysisResult(abc.ABC):
+    ...
+
+
+class Analysis(abc.ABC):
+    @_beartype.beartype
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/_pass.py:214]: docstring

File: torch/onnx/_internal/fx/analysis/unsupported_nodes.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+from torch.onnx._internal.fx import _pass, diagnostics, function_dispatcher
+
+
+@dataclasses.dataclass
+class UnsupportedFxNodesAnalysisResult(_pass.AnalysisResult):
+    unsupported_op_to_target_mapping: Dict[str, Set[str]]
+
+
+class UnsupportedFxNodesAnalysis(_pass.Analysis):
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/analysis/unsupported_nodes.py:14]: Is this already in use?

File: torch/onnx/_internal/fx/function_dispatcher.py [hunk 4]
```diff
@@ -306,3 +309,82 @@ def _create_onnx_friendly_decomposition_table() -> (
...
+    ):
+        for node_arg in node.args:
+            if (not isinstance(node_arg, (torch.fx.Node, int, float))) or (
+                isinstance(node_arg, torch.fx.Node)
+                and not isinstance(node_arg.meta["val"], (torch.SymInt, torch.SymFloat))
+            ):
+                # TODO: reduce number of explicit initializations.
+                # TODO: Log location, stack.
+                diagnostic = diagnostics.UnsupportedFxNodeDiagnostic(
+                    diagnostics.rules.no_symbolic_function_for_call_function,
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/function_dispatcher.py:333]: Not sure if I understand this correctly. This PR registers two rules, but only `no_symbolic_function_for_call_function` is actually guarding errors. It seems the other one is prepared for some kind of full analysis, and it's not done yet? Which means it's not an actual rule to apply on converter code. It's rather a "wrapper rule" on the rules?

File: torch/onnx/_internal/fx/analysis/unsupported_nodes.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/analysis/unsupported_nodes.py:14]: It is. It is invoked before calling `export_fx_to_onnxscript`. It is also how the benchmark gathers all missing op info for a model: by analyzing the diagnostics emitted by this analysis.

Comment (titaiwangms) [reply torch/onnx/_internal/fx/analysis/unsupported_nodes.py:14]: I see

File: torch/onnx/_internal/fx/function_dispatcher.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/function_dispatcher.py:333]: Why do we need another layer to reveal unsupported node?

**Review by titaiwangms — APPROVED** [review]

**Pull request closed without merging**